walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,109 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+
+    // If the requested time is longer than the entire day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+
+    if (attendees.isEmpty() && !request.getOptionalAttendees().isEmpty()) {
+        attendees = request.getOptionalAttendees();
+    }
+
+    // If there are no attendees or there are no events during the day.
+    if (attendees.isEmpty() || events.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Attending list & Optional list
+    List<TimeRange> eventsAL = new ArrayList<>();
+    List<TimeRange> eventsOL = new ArrayList<>();
+
+    // Add events that share attendees.
+    for (Event event : events) {
+        if (!Collections.disjoint(event.getAttendees(), attendees)) {
+            eventsAL.add(event.getWhen());
+        } else if (!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) {
+            eventsOL.add(event.getWhen());
+        }
+    }
+
+    // Check if one event contains another.
+    for (int i = 0; i < eventsAL.size() - 1; i++) {
+        if (eventsAL.get(i).contains(eventsAL.get(i + 1))) {
+            eventsAL.remove(eventsAL.get(i + 1));
+        }
+    }
+
+    List<TimeRange> openSpaces = openSlots(eventsAL);
+    Collection<TimeRange> possibilities = new ArrayList<>();
+
+    for (int i = 0; i < openSpaces.size(); i++) {
+
```

Comment:
Nit: Remove blank line at beginning of for loop body

Reply:
Done!